File: src/416_partition_equal_subset_sum/partition_equal_subset_sum.py
from typing import List


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        numsTotal = sum(nums)
        if numsTotal % 2:
            return False
        targetSum = numsTotal // 2
        dp = [False] * (targetSum + 1)
        dp[0] = True

        for num in nums:
            # Update DP for each number, move backwards to reuse the same number
            for j in range(targetSum, num - 1, -1):
                dp[j] = dp[j] or dp[j - num]

        return dp[targetSum]


# Helper functions for testing


def test_case(nums, expected):
    sol = Solution()
    actual = sol.canPartition(nums)
    print(f"{actual} | {expected}")


# Test cases

if __name__ == "__main__":
    test_case([1, 5, 11, 5], True)
    test_case([1, 2, 3, 5], False)
    test_case([1, 1], True)
    test_case([1], False)
    test_case([0, 0], True)
    test_case([2, 2, 2, 2], True)

File: src/416_partition_equal_subset_sum/PartitionEqualSubsetSum.java
import java.util.*;

class Solution {
    public boolean canPartition(int[] nums) {
        int totalSum = 0;
        for (int num : nums) {
            totalSum += num;
        }
        if (totalSum % 2 != 0) {
            return false;
        }
        int target = totalSum / 2;
        boolean[] dp = new boolean[target + 1];
        dp[0] = true;
        for (int num : nums) {
            for (int j = target; j >= num; j--) {
                dp[j] = dp[j] || dp[j - num];
            }
        }
        return dp[target];
    }
}

class PartitionEqualSubsetSum {
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Test cases
        testCase(sol, new int[]{1,5,11,5}, true);
        testCase(sol, new int[]{1,2,3,5}, false);
        testCase(sol, new int[]{1,1}, true);
        testCase(sol, new int[]{1}, false);
        testCase(sol, new int[]{0,0}, true);
        testCase(sol, new int[]{2,2,2,2}, true);
    }
    
    private static void testCase(Solution sol, int[] nums, boolean expected) {
        boolean actual = sol.canPartition(nums);
        System.out.println(actual + " | " + expected);
    }
}
